File: backend/src/main/java/com/impaintor/feature/realtime/dto/outbound/RoleAssignment.java
package com.impaintor.feature.realtime.dto.outbound;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Asignación de rol enviada una sola vez al inicio de partida vía
 * {@code /user/queue/private}. La forma JSON respeta la sección 6.2 de CLAUDE.md:
 * {@code type:"ROLE_ASSIGNMENT", role:"PAINTER"|"IMPOSTOR", ...}.
 */
public sealed interface RoleAssignment {

    String type();

    String role();

    record Painter(
            @JsonProperty("type") String type,
            @JsonProperty("role") String role,
            @JsonProperty("word") String word
    ) implements RoleAssignment {
        public Painter(String word) {
            this("ROLE_ASSIGNMENT", "PAINTER", word);
        }
    }

    record Impostor(
            @JsonProperty("type") String type,
            @JsonProperty("role") String role,
            @JsonProperty("hint") String hint,
            @JsonProperty("lives") int lives
    ) implements RoleAssignment {
        public Impostor(String hint, int lives) {
            this("ROLE_ASSIGNMENT", "IMPOSTOR", hint, lives);
        }
    }
}

File: frontend/src/app/features/game/models/role-assignment.ts
// Espejo TypeScript del sealed interface RoleAssignment del backend Track D.
// Mantener sincronizado con:
//   backend/src/main/java/com/impaintor/feature/realtime/dto/outbound/RoleAssignment.java

export type Role = 'PAINTER' | 'IMPOSTOR';

export type RoleAssignment =
  | { type: 'ROLE_ASSIGNMENT'; role: 'PAINTER'; word: string }
  | { type: 'ROLE_ASSIGNMENT'; role: 'IMPOSTOR'; hint: string; lives: number };

/** Resultado privado de un intento de adivinación del impostor. */
export interface GuessResult {
  type: 'GUESS_RESULT';
  correct: boolean;
  livesRemaining: number;
}

/** ELO delta sent privately to each player after a game ends. */
export interface EloUpdate {
  type: 'ELO_UPDATE';
  eloChange: number;
}

/** Mensaje privado entrante en /user/queue/private — union de los anteriores. */
export type PrivateMessage = RoleAssignment | GuessResult | EloUpdate;
